File: experiments/complexity_analysis/On3_cubic/slow_solutions/slow_solution.py
def inefficient_matrix_multiply(A, B):
    """Algorithmically equivalent but inefficient: O(n⁴) using redundant computations"""
    # Each dot product is computed n times and averaged, maintaining mathematical equivalence
    n = len(A)
    C = [[0 for _ in range(n)] for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            dot_product_sum = 0
            # Compute dot product n times (inefficient)
            for repetition in range(n):
                for k in range(n):
                    dot_product_sum += A[i][k] * B[k][j]
            # Divide by n to get back original dot product
            C[i][j] = dot_product_sum // n
    
    return C

# Read input
n = int(input())

# Read matrix A
A = []
for i in range(n):
    row = list(map(int, input().split()))
    A.append(row)

# Read matrix B
B = []
for i in range(n):
    row = list(map(int, input().split()))
    B.append(row)

# Calculate C = A × B using inefficient approach
C = inefficient_matrix_multiply(A, B)

# Calculate sum of all elements in result matrix
matrix_sum = sum(sum(row) for row in C)

# Output sum
print(matrix_sum)

File: experiments/complexity_analysis/On3_cubic/slow_solutions/slow_solution.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<vector<long long>> inefficientMatrixMultiply(const vector<vector<long long>>& A, const vector<vector<long long>>& B) {
    // Algorithmically equivalent but inefficient: O(n⁴) using redundant computations
    // Each dot product is computed n times and averaged, maintaining mathematical equivalence
    int n = A.size();
    vector<vector<long long>> C(n, vector<long long>(n, 0));
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            long long dot_product_sum = 0;
            // Compute dot product n times (inefficient)
            for (int repetition = 0; repetition < n; repetition++) {
                for (int k = 0; k < n; k++) {
                    dot_product_sum += A[i][k] * B[k][j];
                }
            }
            // Divide by n to get back original dot product
            C[i][j] = dot_product_sum / n;
        }
    }
    return C;
}

int main() {
    int n;
    cin >> n;
    
    // Read matrix A
    vector<vector<long long>> A(n, vector<long long>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> A[i][j];
        }
    }
    
    // Read matrix B
    vector<vector<long long>> B(n, vector<long long>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> B[i][j];
        }
    }
    
    // Calculate C = A × B using inefficient approach
    vector<vector<long long>> C = inefficientMatrixMultiply(A, B);
    
    // Calculate sum of all elements in result matrix
    long long sum = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            sum += C[i][j];
        }
    }
    
    cout << sum << endl;
    
    return 0;
}
